pkg/mcs/tso/server/config.go

```diff
@@ -0,0 +1,254 @@
+			return err
+		}
+	}
+
+	// Ignore the error check here
+	configutil.AdjustCommandlineString(flagSet, &c.Log.Level, "log-level")
+	configutil.AdjustCommandlineString(flagSet, &c.Log.File.Filename, "log-file")
+	configutil.AdjustCommandlineString(flagSet, &c.Metric.PushAddress, "metrics-addr")
+	configutil.AdjustCommandlineString(flagSet, &c.Security.CAPath, "cacert")
+	configutil.AdjustCommandlineString(flagSet, &c.Security.CertPath, "cert")
+	configutil.AdjustCommandlineString(flagSet, &c.Security.KeyPath, "key")
+	configutil.AdjustCommandlineString(flagSet, &c.BackendEndpoints, "backend-endpoints")
+	configutil.AdjustCommandlineString(flagSet, &c.ListenAddr, "listen-addr")
+
+	return c.Adjust(meta, false)
+}
+
+// Adjust is used to adjust the TSO configurations.
+func (c *Config) Adjust(meta *toml.MetaData, reloading bool) error {
+	configMetaData := configutil.NewConfigMetadata(meta)
+	if err := configMetaData.CheckUndecoded(); err != nil {
+		c.WarningMsgs = append(c.WarningMsgs, err.Error())
+	}
+	if c.Name == "" {
+		hostname, err := os.Hostname()
+		if err != nil {
+			return err
+		}
+		configutil.AdjustString(&c.Name, fmt.Sprintf("%s-%s", defaultName, hostname))
+	}
+	configutil.AdjustString(&c.DataDir, fmt.Sprintf("default-datadir.%s", c.Name))
+	adjustPath(&c.DataDir)
+
+	if err := c.Validate(); err != nil {
+		return err
+	}
+
+	configutil.AdjustString(&c.BackendEndpoints, defaultBackendEndpoints)
+	configutil.AdjustString(&c.ListenAddr, defaultListenAddr)
+
+	configutil.AdjustDuration(&c.MaxResetTSGap, defaultMaxResetTSGap)
+	configutil.AdjustInt64(&c.LeaderLease, defaultLeaderLease)
+	configutil.AdjustDuration(&c.TSOSaveInterval, defaultTSOSaveInterval)
+	configutil.AdjustDuration(&c.TSOUpdatePhysicalInterval, defaultTSOUpdatePhysicalInterval)
+
+	if c.TSOUpdatePhysicalInterval.Duration > maxTSOUpdatePhysicalInterval {
+		c.TSOUpdatePhysicalInterval.Duration = maxTSOUpdatePhysicalInterval
+	} else if c.TSOUpdatePhysicalInterval.Duration < minTSOUpdatePhysicalInterval {
+		c.TSOUpdatePhysicalInterval.Duration = minTSOUpdatePhysicalInterval
+	}
+	if c.TSOUpdatePhysicalInterval.Duration != defaultTSOUpdatePhysicalInterval {
+		log.Warn("tso update physical interval is non-default",
+			zap.Duration("update-physical-interval", c.TSOUpdatePhysicalInterval.Duration))
+	}
+
+	if !configMetaData.IsDefined("enable-grpc-gateway") {
+		c.EnableGRPCGateway = defaultEnableGRPCGateway
+	}
+
+	c.adjustLog(configMetaData.Child("log"))
+	c.Security.Encryption.Adjust()
+
+	if len(c.Log.Format) == 0 {
+		c.Log.Format = defaultLogFormat
+	}
+
+	return nil
+}
+
+func adjustPath(p *string) {
+	absPath, err := filepath.Abs(*p)
+	if err == nil {
+		*p = absPath
+	}
+}
+
+func (c *Config) adjustLog(meta *configutil.ConfigMetaData) {
+	if !meta.IsDefined("disable-error-verbose") {
+		c.Log.DisableErrorVerbose = defaultDisableErrorVerbose
+	}
+}
+
+// Validate is used to validate if some configurations are right.
+func (c *Config) Validate() error {
```

Comment:
Do we need to check TSOSaveInterval and TSOUpdatePhysicalInterval

Reply:
it's checked and adjusted in Adjust()